quickwit/quickwit-control-plane/src/scheduler.rs

```diff
@@ -17,37 +17,585 @@
+                .await
+            {
+                Ok(mut indexing_client) => {
+                    if let Err(error) = indexing_client
+                        .apply_indexing_plan(ApplyIndexingPlanRequest {
+                            indexing_tasks: indexing_tasks.clone(),
+                        })
+                        .await
+                    {
+                        error!(indexer_node_id=%indexer.node_id, err=?error, "Error occured when appling indexing plan to indexer.");
+                    }
+                }
+                Err(error) => {
+                    error!(indexer_node_id=%indexer.node_id, err=?error,
+                        "Cannot create indexing service client for indexer, skip indexing plan.",
+                    );
+                }
+            }
+        }
+        self.state.num_applied_physical_indexing_plan += 1;
+        self.state.last_applied_plan_timestamp = Some(OffsetDateTime::now_utc().unix_timestamp());
+        self.state.last_applied_physical_plan = Some(new_physical_plan);
+    }
+}
+
+#[async_trait]
+impl Handler<NotifyIndexChangeRequest> for IndexingScheduler {
+    type Reply = ();
+
+    async fn handle(
+        &mut self,
+        _: NotifyIndexChangeRequest,
+        _: &ActorContext<Self>,
+    ) -> Result<(), ActorExitStatus> {
+        info!("Index change notification: shedule indexing plan.");
+        if let Err(error) = self.schedule_indexing_plan().await {
+            error!("Error when scheduling indexing plan: `{}`.", error);
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct ControlPlanLoop;
+
+#[async_trait]
+impl Handler<ControlPlanLoop> for IndexingScheduler {
+    type Reply = ();
+
+    async fn handle(
+        &mut self,
+        _message: ControlPlanLoop,
+        ctx: &ActorContext<Self>,
+    ) -> Result<(), ActorExitStatus> {
+        if let Err(error) = self.control_running_plan().await {
+            error!("Error when controlling the running plan: `{}`.", error);
+        }
+        ctx.schedule_self_msg(HEARTBEAT, ControlPlanLoop).await;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct RefreshPlanLoop;
+
+#[async_trait]
+impl Handler<RefreshPlanLoop> for IndexingScheduler {
+    type Reply = ();
+
+    async fn handle(
+        &mut self,
+        _message: RefreshPlanLoop,
+        ctx: &ActorContext<Self>,
+    ) -> Result<(), ActorExitStatus> {
+        if let Err(error) = self.schedule_indexing_plan().await {
+            error!("Error when scheduling indexing plan: `{}`.", error);
+        }
+        ctx.schedule_self_msg(REFRESH_PLAN_LOOP_INTERVAL, RefreshPlanLoop)
+            .await;
+        Ok(())
+    }
+}
+
+/// Returns true if equal, if not, logs the difference between the plans.
+/// The plans are considered equal if they have the same node IDs and if
+/// each node has the same running tasks.
+fn are_indexing_plans_equal(
+    running_plan: &HashMap<String, Vec<IndexingTask>>,
```

Comment:
Can you detail how we got the running_plan? Is it surfaced by chitchat?

Reply:
Yes, it comes from chitchat, just added it in the docs.